src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopeTracker.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyInjection.ServiceLookup;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    internal class ScopeTracker
+    {
+        public State Allocate() => new State(this);
+
+        private void Track(State state)
+        {
+            DependencyInjectionEventSource.Log.ScopeDisposed(this, state);
+        }
+
+        public class State
+        {
+            private readonly ScopeTracker _tracker;
+
+            public IDictionary<ServiceCacheKey, object> ResolvedServices { get; }
+            public List<object> Disposables { get; set; }
+
+            public int DisposableServicesCount => Disposables?.Count ?? 0;
+            public int ResolvedServicesCount => ((Dictionary<ServiceCacheKey, object>)ResolvedServices).Count;
```

Comment:
Is this cast safe? Or do we only expect `ResolvedServicesCount` to be called on non-root scopes?

Reply:
IT used to be safe, but now it's not.

Reply:
Removed cast